### PR TITLE
Feature 1528 plot_point_obs

### LINKED ISSUE #213: Enhance plot_point_obs to filter observations based on the quality flag value in the obs_qty() variable.
Enhance plot_point_obs to filter observations based on the quality flag value in the obs_qty() variable.  Probably just add a new command line argument.
<i>[MET-213] created by johnhg</i>

### LINKED ISSUE #1052: Enhance plot_point_obs to color the points by message type and/or variable name.
This request came up during the AF project.  Would like to enhance the plotting options for plot_point_obs.  See Jonathan Vigh who already did this using NCL and implement the options he chose.  One tricky part is adding a legend for this.
<br/>

<br/>
Consider supporting different colors and/or plotting symbols.
<br/>

<i>[MET-1052] created by johnhg</i>

### LINKED ISSUE #1528: Consider making the plotting options of plot_point_obs more configurable.
## Describe the New Feature ##

There are many ways that plot_point_obs could be improved. Listed below are some recommended changes based on a recent met_help question from Binyu:

We should change plot_point_obs’s command line argument from:
-data_file path to gridded data file
To:
-plot_grid filename, grid name, grid spec

And then just reuse the same logic for parsing “to_grid” in regrid_data_plane. Still quietly support -data_file.

From binyu’s met_help question. She also want to filter obs by their height or pressure.

We could also add an optional plot_point_obs config file consisting of a “plot_options” array where each element is a dictionary defining a filter and set of plotting options. The logic would be... for each plot_options entry, loop through all of the input point observations and plot the matching dots. The alternative logic would be to only read the obs once, and then plot a dot using the matching plot_options entry. Or if none match, plot nothing. I think the second logic is simpler.

// Default config file

outline_color = “”;
fill_color = “red”;
fill_color_table = “”; // takes precedence over fill_color. Does this mean we need a color bar?
dotsize = 10; // could be a function of obs value?

message_type = [];
obs_var = [];
height_thresh = NA;
pressure_thresh = NA;
obs_thresh = NA;
obs_qc = [];
// what about convert and censor_thresh/censor_val?

plot_options = [
  { fill_color = “red”; }
];

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

### Time Estimate ###
3 days.

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [ ] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Pull Request Testing ##

This is a large overhaul of the plot_point_obs utility that adds a configuration file and makes it highly configurable.
Note that this PR fixes 3 GitHub issues #1528, #1052, and #213.

@jprestop: Please review the usage statement, config file options, and documentation updates.
@michelleharrold and @TaraJensen: Please test out the tool and let me know if any changes to the functionality are needed.

- [x] Describe testing already performed for these changes:</br>
Added a new unit tests and confirmed that all of the tests in unit_plot_point_obs.xml run. The 2 existing tests should produce the same visual output as before, although the contents of the PostScript file may differ slight. The tool adjusts the margin sizes only if the user has requested that a colorbar be added.

Compiled the code for this PR on kiowa in /d1/projects/MET/MET_pull_requests/met-10.0_beta2/feature_1528.
Ran the full set of unit tests and will diffs the output against the develop branch.

- [x] Recommend testing for the reviewer to perform, including the location of input datasets:</br>
Play around with the new plotting options to make sure they work as expected. Review the updates to the plotting chapter about the plot_point_obs tool. Review the addition to unit_plot_point_obs.xml to make sure the plot makes sense scientifically and exercises a reasonable set of config file options.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
New image named "nam_and_ndas.20120409.t12z.prepbufr_CONFIG.ps" is created by unit_plot_point_obs.xml

- [x] After merging, should the reviewer **DELETE** the feature branch from GitHub? **[No]**</br>
This is a big change and the reviews may have requested changes. Please leave the branch until all those are addressed.

## Pull Request Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, and **Milestone**
- [x] After submitting the PR, select **Linked Issues** with the original issue number.

### CODE REVIEW

File: met/data/config/PlotPointObsConfig_default [hunk 1]
```diff
@@ -0,0 +1,72 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Plot-Point-Obs configuration file.
+//
+// For additional information, see the MET_BASE/config/README file.
```
Comment (jprestop) [inline met/data/config/PlotPointObsConfig_default]:
```suggestion
For additional information, see the README file in the MET User's Guide.
```
Please update the comment in the README file.

**Review by jprestop — APPROVED** [review]
I reviewed the usage statement, config file options, and documentation updates.  I made two suggested changes.  I also noticed that the files in test/config should no longer say "// For additional information, see the MET_BASE/config/README file." but rather should say "// For additional information, see the README file in the MET User's Guide.", so I will fix those up. Thank you for all of your work on this!

File: test/config/PlotPointObsConfig [hunk 2]
```diff
@@ -0,0 +1,96 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Plot-Point-Obs configuration file.
+//
+// For additional information, see the MET_BASE/config/README file.
```
Comment (jprestop) [inline test/config/PlotPointObsConfig]:
```suggestion
// For additional information, see the README file in the MET User's Guide.
```

**Review by jprestop — APPROVED** [review]

**Comment by JohnHalleyGotway** [comment]
Ran the unit tests and diffed the output with develop-ref. As expected, there is 1 new plot_point_obs output file and 3 existing ones whose contents have changed slightly. However, a visual inspection reveals that the images look the same.

On kiowa in:
/d1/projects/MET/MET_pull_requests/met-10.0_beta2/feature_1528/MET-feature_1528_plot_point_obs_into_develop

file1: test_output/met_test_scripts/plot_point_obs/sample_pb.ps
file2: /d1/projects/MET/MET_regression/develop/NB20201116/MET-develop-ref/test_output/met_test_scripts/plot_point_obs/sample_pb.ps

file1: test_output/plot_point_obs/gdas1.20120409.t12z.prepbufr_TMP_ADPUPA.ps
file2: /d1/projects/MET/MET_regression/develop/NB20201116/MET-develop-ref/test_output/plot_point_obs/gdas1.20120409.t12z.prepbufr_TMP_ADPUPA.ps

file1: test_output/plot_point_obs/ndas.20120409.t12z.prepbufr.tm00.nr_G218.ps
file2: /d1/projects/MET/MET_regression/develop/NB20201116/MET-develop-ref/test_output/plot_point_obs/ndas.20120409.t12z.prepbufr.tm00.nr_G218.ps

See attached tarfile with PNG versions of these images.
[feature_1528_png.tar.gz](https://github.com/dtcenter/MET/files/5550290/feature_1528_png.tar.gz)

**Comment by JohnHalleyGotway** [comment]
Please untar attached file and open _build/html/index.html in a web browser (and keep your fingers crossed).

[docs_html.tar.gz](https://github.com/dtcenter/MET/files/5550334/docs_html.tar.gz)

The plot_point_obs doc updates are in _build/html/Users_Guide/plotting.html

**Comment by michelleharrold** [comment]
Tested new plot_plot_obs functionality on kiowa with GDAS prepbufr file as well as AF grib2 output. All testing permutations were successful. Some notes on testing:

1) The usage statement should be updated to include the -data_file option.
2) Two "nice-to-have's" but not necessary: 1) Having a 'title' option similar to functionality in plot_data_plane, and 2) if the AF is interested in plotting model terrain, it might be nice to include a terrain-specific color table in the colortables directory

**Review by michelleharrold — APPROVED** [review]

**Comment by JohnHalleyGotway** [comment]
Michelle, thanks for the feedback.

The "-title" command line option actually already exists, but I failed to list it in the usage statement. I'll add that into the usage statement and update the docs.

The "-plot_grid" command line option expands and replaces the "-data_file" option. It can now be specified as the path to a gridded data file (old), named grid (new), or grid specification string (new). But I kept in quiet support for the "-data_file" option for backwards compatibility. Does that seem like a good approach?

The files in data/config/colortables/NCL_colortables were derived from:
https://www.ncl.ucar.edu/Document/Graphics/color_table_gallery.shtml
MET currently has 56 NCL colortables but it looks like there are several new ones, including a few with "topo" in the name. Should I pull the latest set of NCL colortables from...
https://github.com/NCAR/ncl/tree/6.6.2/ni/src/db/colormaps
There are 270 *.rgb files with a total size of 1.2M.

**Pull request merged**